plugins/map_matching.hpp

```diff
@@ -0,0 +1,314 @@
+
+#include "plugin_base.hpp"
+
+#include "../algorithms/bayes_classifier.hpp"
+#include "../algorithms/object_encoder.hpp"
+#include "../data_structures/search_engine.hpp"
+#include "../descriptors/descriptor_base.hpp"
+#include "../descriptors/json_descriptor.hpp"
+#include "../routing_algorithms/map_matching.hpp"
+#include "../util/compute_angle.hpp"
+#include "../util/integer_range.hpp"
+#include "../util/simple_logger.hpp"
+#include "../util/json_logger.hpp"
+#include "../util/json_util.hpp"
+#include "../util/string_util.hpp"
+
+#include <cstdlib>
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+template <class DataFacadeT> class MapMatchingPlugin : public BasePlugin
+{
+  private:
+    std::shared_ptr<SearchEngine<DataFacadeT>> search_engine_ptr;
+
+    using ClassifierT = BayesClassifier<LaplaceDistribution, LaplaceDistribution, double>;
+    using TraceClassification = ClassifierT::ClassificationT;
+
+  public:
+    MapMatchingPlugin(DataFacadeT *facade, const int max_locations_map_matching)
+    : descriptor_string("match")
+    , facade(facade)
+    , max_locations_map_matching(max_locations_map_matching)
+    // the values where derived from fitting a laplace distribution
+    // to the values of manually classified traces
+    , classifier(LaplaceDistribution(0.0057154021891018675, 0.020294704891166186),
+                 LaplaceDistribution(0.11467696742821254, 0.49918444000368756),
+                 0.7977883096366508) // valid apriori probability
+    {
+        search_engine_ptr = std::make_shared<SearchEngine<DataFacadeT>>(facade);
+    }
+
+    virtual ~MapMatchingPlugin() { }
+
+    const std::string GetDescriptor() const final { return descriptor_string; }
+
+    TraceClassification classify(float trace_length, float matched_length, int removed_points) const
+    {
+        double distance_feature = -std::log(trace_length) + std::log(matched_length);
+
+        // matched to the same point
+        if (!std::isfinite(distance_feature))
+        {
+            return std::make_pair(ClassifierT::ClassLabel::NEGATIVE, 1.0);
+        }
+
+        auto label_with_confidence =  classifier.classify(distance_feature);
+
+        // "second stage classifier": if we need to remove points there is something fishy
+        if (removed_points > 0)
+        {
+            return std::make_pair(ClassifierT::ClassLabel::NEGATIVE, 1.0);
+        }
+
+        return label_with_confidence;
+    }
+
+    bool getCandiates(const std::vector<FixedPointCoordinate>& input_coords, std::vector<double>& sub_trace_lengths, Matching::CandidateLists& candidates_lists)
+    {
+        double last_distance = coordinate_calculation::great_circle_distance(
+            input_coords[0],
+            input_coords[1]);
+        sub_trace_lengths.resize(input_coords.size());
+        sub_trace_lengths[0] = 0;
+        for (const auto current_coordinate : osrm::irange<std::size_t>(0, input_coords.size()))
```

Comment:
`osrm::irange(0u, input_coords.size())` should work to make the code more readable.